esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+// Licensed to Elasticsearch B.V. under one or more agreements.
+// Elasticsearch B.V. licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+package esutil
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/elastic/go-elasticsearch/v8"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+// BulkIndexer represents a parallel, asynchronous, efficient indexer for Elasticsearch.
+//
+type BulkIndexer interface {
+	// Add adds an item to the indexer. It returns an error when the item cannot be added.
+	// Use the OnSuccess and OnFailure callbacks to get the operation result for the item.
+	//
+	// You must call the Close() method after you're done adding items.
+	//
+	// It is safe for concurrent use. When it's called from goroutines,
+	// they must finish before the call to Close, eg. using sync.WaitGroup.
+	Add(context.Context, BulkIndexerItem) error
```

Comment:
Did you consider exposing the channel, so that users could select on it? e.g.

```go
type BulkIndexer struct {
    C() chan<- BulkIndexerItem
    Close(context.Context) error
    Stats() BulkIndexerStats
}
```

Reply:
Fair enough, and I agree - just wondering if you considered this option, as I brought it up elsewhere. I think selecting on add would _probably_ be unusual anyway, so it isn't so relevant here.

Reply:
💯 for not using go-channels. I can always wrap the Indexer in a go-routine and use channels for the callback, but working around APIs that enforce go-channels can be a pain.